modules/libcom/RTEMS/posix/rtems_init.c

```diff
@@ -0,0 +1,1159 @@
+static const iocshFuncDef zonesetFuncDef = {"zoneset",1,zonesetArgs
+#ifdef IOCSHFUNCDEF_HAS_USAGE
+                                           , "set timezone (obsolete?)"
+#endif
+                                           };
+static void zonesetCallFunc(const iocshArgBuf *args)
+{
+    zoneset(args[0].sval);
+}
+
+
+/*
+ * Register RTEMS-specific commands
+ */
+static void iocshRegisterRTEMS (void)
+{
+    iocshRegister(&netStatFuncDef, netStatCallFunc);
+    iocshRegister(&heapSpaceFuncDef, heapSpaceCallFunc);
+#ifndef OMIT_NFS_SUPPORT
+    iocshRegister(&nfsMountFuncDef, nfsMountCallFunc);
+#endif
+    iocshRegister(&zonesetFuncDef, &zonesetCallFunc);
+    iocshRegister(&rtshellFuncDef, &rtshellCallFunc);
+#if __RTEMS_MAJOR__ > 4
+    rtems_shell_init_environment();
+#endif
+}
+
+/*
+ * Set up the console serial line (no handshaking)
+ */
+static void
+initConsole (void)
+{
+    struct termios t;
+
+printf("\n initConsole --- Info ---\n");
+printf("stdin: fileno: %d, ttyname: %s\n", fileno(stdin), ttyname(fileno(stdin)));
+printf("stdout: fileno: %d, ttyname: %s\n", fileno(stdout), ttyname(fileno(stdout)));
+printf("stderr: fileno: %d, ttyname: %s\n", fileno(stderr), ttyname(fileno(stderr)));
+
+    if (tcgetattr (fileno (stdin), &t) < 0) {
+        printf ("tcgetattr failed: %s\n", strerror (errno));
+        return;
+    }
+    t.c_iflag &= ~(IXOFF | IXON | IXANY);
+    if (tcsetattr (fileno (stdin), TCSANOW, &t) < 0) {
+        printf ("tcsetattr failed: %s\n", strerror (errno));
+        return;
+    }
+}
+
+/*
+ * Hook to ensure that BSP cleanup code gets run on exit
+ */
+static void
+exitHandler(void)
+{
+    rtems_shutdown_executive(0);
+}
+
+#ifndef RTEMS_LEGACY_STACK
+static char* getPrimaryNetworkInterface(void)
+{
+    // lookup available network interfaces
+    char ifnamebuf[IF_NAMESIZE];
+    char *ifname;
+    // get primary network interface
+    ifname = if_indextoname(1, &ifnamebuf[0]);
+    if (ifname == NULL) return (NULL);
+    printf("\n***** Primary Network interface : %s *****\n", ifname);
+    return (ifname);
+}
+
+static void
+dhcpcd_hook_handler(rtems_dhcpcd_hook *hook, char *const *env)
+{
+    int bound = 0;
+    char iName[16];
+    char *name;
+    char *value;
+
+	(void)hook;
+
+    char ifnamebuf[IF_NAMESIZE];
+    sprintf(ifnamebuf, "%s", getPrimaryNetworkInterface());
+
+    while (*env != NULL) {
+        name = strtok(*env,"=");
+        value = strtok(NULL,"=");
+        printf("all out ---> %s = %s\n", name, value);
+        if (!strncmp(name, "interface", 9) && !strcmp(value,  ifnamebuf))
+			 strncpy(iName, value, 16);
+		if (!strncmp(name, "reason", 6) && !strncmp(value, "BOUND", 5)){
+            printf ("Interface %s bounded\n", iName);
+            bound = 1;
+		}
+		if (bound) {
+            // as there is no ntp-support in rtems-libbsd, we call our own client
+            if(!strncmp(name, "new_ntp_servers", 15))
+                strcpy(rtemsInit_NTP_server_ip,value);
```

Comment:
Some use of `sizeof(rtemsInit_NTP_server_ip)` would be prudent.

Reply:
I changed a lot of this string handling to use sizeof() to check sizes, use snprintf() so that we can get reliable null termination (vs. strncpy), and use strtok_r rather than strtok.